indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java

```diff
@@ -0,0 +1,636 @@
+        + "\nExclude: \n" + prefixExcludeRules + "\n" + regexExcludeRules;
+  }
+
+  /**
+   * Builds an IncludeExcludeFilter instance from configured properties. With no
+   * properties, returns a filter that allows everything.
+   *
+   * @return an IncludeExcludeFilter
+   */
+  public static IncludeExcludeFilter fromConfiguration() {
+    checkState(Configuration.isInitialized());
+    Set<String> filterProperties = Configuration.getConfig()
+        .stringPropertyNames()
+        .stream()
+        .filter(key -> key.startsWith(FILTER_CONFIG_PREFIX))
+        .collect(Collectors.toSet());
+    if (filterProperties.isEmpty()) {
+      return new IncludeExcludeFilter(ImmutableList.of());
+    }
+    HashMap<String, Rule.Builder> ruleBuilders = new HashMap<>();
+    for (String propertyName : filterProperties) {
+      String[] propertyNameParts = propertyName.split("\\.");
+      if (propertyNameParts.length != 3) {
+        throw new InvalidConfigurationException("Unknown property " + propertyName);
+      }
+      String ruleName = propertyNameParts[1];
+      String rulePropertyType = propertyNameParts[2];
+      String propertyValue = Configuration.getString(propertyName, null).get();
+      Rule.Builder builder = ruleBuilders.get(ruleName);
+      if (builder == null) {
+        builder = new Rule.Builder(ruleName);
+        ruleBuilders.put(ruleName, builder);
+      }
+      if (rulePropertyType.equals("itemType")) {
+        builder.setItemType(propertyValue);
+      } else if (rulePropertyType.equals("filterType")) {
+        builder.setFilterType(propertyValue);
+      } else if (rulePropertyType.equals("filterPattern")) {
+        builder.setFilterPattern(propertyValue);
+      } else if (rulePropertyType.equals("action")) {
+        builder.setAction(propertyValue);
+      } else {
+        throw new InvalidConfigurationException("Unknown property " + propertyName);
+      }
+    }
+    List<Rule> rules = new ArrayList<>();
+    for (Rule.Builder builder : ruleBuilders.values()) {
+      try {
+        rules.add(builder.build());
+      } catch (IllegalStateException | IllegalArgumentException e) {
+        throw new InvalidConfigurationException(e.getMessage());
+      }
+    }
+    return new IncludeExcludeFilter(rules);
+  }
+
+  /**
+   * Returns true if the given string is included based on the configured include/exclude
+   * patterns.
+   *
+   * @param value a value to test
+   * @param itemType the type of item; values of {@code VIRTUAL_CONTAINER_ITEM} will be
+   * mapped to {@code CONTAINER_ITEM}
+   * @return true if the value is included based on the configuration
+   */
+  public boolean isAllowed(String value, ItemType itemType) {
+    if (itemType.equals(ItemType.VIRTUAL_CONTAINER_ITEM)) {
+      itemType = ItemType.CONTAINER_ITEM;
+    }
+    List<Rule> excludeRules =
+        Stream.concat(prefixExcludeRules.stream(), regexExcludeRules.get(itemType).stream())
+        .collect(Collectors.toList());
```

Comment:
I think that if you collect this to ImmutableList, then evaluateRules can take an ImmutableList.